hazelcast/src/main/java/com/hazelcast/sql/impl/QueryId.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl;
+
+import com.hazelcast.internal.util.UuidUtil;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+
+import java.io.IOException;
+import java.util.UUID;
+
+/**
+ * Cluster-wide unique query ID.
+ */
+public final class QueryId implements IdentifiedDataSerializable {
+    /** Member ID: most significant bits */
+    private long memberIdHigh;
+
+    /** Member ID: least significant bits. */
+    private long memberIdLow;
+
+    /** Local ID: most significant bits. */
+    private long localHigh;
+
+    /** Local ID: least significant bits. */
+    private long localLow;
+
+    public QueryId() {
+        // No-op.
+    }
+
+    QueryId(long memberIdHigh, long memberIdLow, long localHigh, long localLow) {
+        this.memberIdHigh = memberIdHigh;
+        this.memberIdLow = memberIdLow;
+        this.localHigh = localHigh;
+        this.localLow = localLow;
+    }
+
+    /**
+     * Create new query ID for the given member.
+     *
+     * @param memberId Member ID.
+     * @return Query ID.
+     */
+    public static QueryId create(UUID memberId) {
+        UUID qryId = UuidUtil.newUnsecureUUID();
+
+        return new QueryId(
+            memberId.getMostSignificantBits(),
+            memberId.getLeastSignificantBits(),
+            qryId.getMostSignificantBits(),
+            qryId.getLeastSignificantBits()
+        );
+    }
+
+    public UUID getMemberId() {
+        return new UUID(memberIdHigh, memberIdLow);
```

Comment:
Why would you create new `UUID` object on every method call? It seems better to create the object once.

Reply:
This method is not used on the hot code paths, mostly in periodic query state checks. So produced garbage should have little to no impact. On the other hand, this saves a little heap space.

So, basically, both approaches - with UUID instance inside the class, or with UUID instance created lazily - are minor micro-optimizations, with no significant side effects.